src/hotspot/share/opto/loopnode.cpp

```diff
@@ -4721,102 +4810,124 @@ void PhaseIdealLoop::verify_compare( Node *n, const PhaseIdealLoop *loop_verify,
+        // Irreducible loops can pick a different header (one of its entries).
+      } else {
+        tty->print_cr("We have loop that verify does not have");
+        child->dump();
+        fail = true;
+      }
+      i++; // step for this
+    } else if (child == nullptr) {
+      if (child_verify->_irreducible && Compile::current()->major_progress()) {
+        // Irreducible loops can pick a different header (one of its entries).
+      } else if (child_verify->_head->as_Region()->is_in_infinite_subgraph()) {
+        // Infinite loops do not get attached to the loop-tree on their first visit.
+        // "this" runs before "loop_verify". It is thus possible that we find the
+        // infinite loop only for "child_verify". Only finding it with "child" would
+        // mean that we lost it, which is not ok.
+      } else {
+        tty->print_cr("Verify has loop that we do not have");
+        child_verify->dump();
+        fail = true;
+      }
+      j++; // step for verify
+    } else {
+      assert(child->_head == child_verify->_head, "We have both and they are equal");
+      fail |= child->verify_tree(child_verify); // Recursion
+      i++; // step for this
+      j++; // step for verify
+    }
+  }
+
+  // Broken part of VerifyLoopOptimizations (D)
+  // Reason:
+  //   split_if has to update the _tail, if it is modified. But that is done by
+  //   checking to what loop the iff belongs to. That info can be wrong, and then
+  //   we do not update the _tail correctly.
+  /*
   Node *tail = _tail;           // Inline a non-updating version of
   while( !tail->in(0) )         // the 'tail()' call.
     tail = tail->in(1);
   assert( tail == loop->_tail, "mismatched loop tail" );
+  */
 
-  // Counted loops that are guarded should be able to find their guards
-  if( _head->is_CountedLoop() && _head->as_CountedLoop()->is_main_loop() ) {
+  if (_head->is_CountedLoop()) {
     CountedLoopNode *cl = _head->as_CountedLoop();
-    Node *init = cl->init_trip();
-    Node *ctrl = cl->in(LoopNode::EntryControl);
-    assert( ctrl->Opcode() == Op_IfTrue || ctrl->Opcode() == Op_IfFalse, "" );
-    Node *iff  = ctrl->in(0);
-    assert( iff->Opcode() == Op_If, "" );
-    Node *bol  = iff->in(1);
-    assert( bol->Opcode() == Op_Bool, "" );
-    Node *cmp  = bol->in(1);
-    assert( cmp->Opcode() == Op_CmpI, "" );
-    Node *add  = cmp->in(1);
-    Node *opaq;
-    if( add->Opcode() == Op_Opaque1 ) {
-      opaq = add;
-    } else {
-      assert( add->Opcode() == Op_AddI || add->Opcode() == Op_ConI , "" );
-      assert( add == init, "" );
-      opaq = cmp->in(2);
-    }
-    assert( opaq->Opcode() == Op_Opaque1, "" );
 
+    Node* ctrl     = cl->init_control();
+    Node* back     = cl->back_control();
+    assert(ctrl != nullptr && ctrl->is_CFG(), "sane loop in-ctrl");
+    assert(back != nullptr && back->is_CFG(), "sane loop backedge");
+    Node* loopexit = cl->loopexit(); // assert implied
```

Comment:
Local variable assignment can be removed.
```suggestion
    cl->loopexit(); // assert implied
```